contracts/lockdrop/src/state.rs

```diff
@@ -14,13 +18,97 @@ pub const ASSET_POOLS: Map<&Addr, PoolInfo> = Map::new("LiquidityPools");
 /// Key is an user address
 pub const USER_INFO: Map<&Addr, UserInfo> = Map::new("users");
 /// Key consists of an Terraswap LP token address, an user address, and a duration
-pub const LOCKUP_INFO: Map<(&Addr, &Addr, U64Key), LockupInfo> = Map::new("lockup_position");
+pub const LOCKUP_INFO: Map<(&Addr, &Addr, U64Key), LockupInfoV2> = Map::new("lockup_position");
+/// Old LOCKUP_INFO storage interface for backward compatibility
+pub const OLD_LOCKUP_INFO: Map<(&Addr, &Addr, U64Key), LockupInfoV1> = Map::new("lockup_position");
 /// Total received asset reward by lockdrop contract per lp token share
 pub const TOTAL_ASSET_REWARD_INDEX: Map<&Addr, Decimal256> = Map::new("total_asset_reward_index");
 /// Last used total asset reward index for user claim ( lp_addr -> user -> duration )
 pub const USERS_ASSET_REWARD_INDEX: Map<(&Addr, &Addr, U64Key), Decimal256> =
     Map::new("users_asset_reward_index");
 
+pub trait CompatibleLoader<K, R> {
+    fn compatible_load(&self, deps: Deps, key: K, generator: &Option<Addr>) -> StdResult<R>;
+
+    fn compatible_may_load(
+        &self,
+        deps: Deps,
+        key: K,
+        generator: &Option<Addr>,
+    ) -> StdResult<Option<R>>;
+}
+
+impl CompatibleLoader<(&Addr, &Addr, U64Key), LockupInfoV2>
+    for Map<'_, (&Addr, &Addr, U64Key), LockupInfoV2>
+{
+    fn compatible_load(
+        &self,
+        deps: Deps,
+        key: (&Addr, &Addr, U64Key),
+        generator: &Option<Addr>,
+    ) -> StdResult<LockupInfoV2> {
+        self.load(deps.storage, key.clone()).or_else(|_| {
+            let old_lockup_info = OLD_LOCKUP_INFO.load(deps.storage, key.clone())?;
+            let mut generator_proxy_debt = Vec::default();
+
+            if generator.is_some() && !old_lockup_info.generator_proxy_debt.is_zero() {
+                let asset = ASSET_POOLS.load(deps.storage, key.0)?;
+                let astro_lp = asset
+                    .migration_info
+                    .expect("Pool should be migrated!")
+                    .astroport_lp_token;
+                let pool_info: PoolInfoResponse = deps.querier.query_wasm_smart(
+                    generator.as_ref().expect("Generator should be set!"),
+                    &GenQueryMsg::PoolInfo {
+                        lp_token: astro_lp.to_string(),
+                    },
+                )?;
+                if let Some((proxy, _)) = pool_info.accumulated_proxy_rewards_per_share.first() {
+                    let proxy_cfg: astroport::generator_proxy::ConfigResponse =
+                        deps.querier.query_wasm_smart(
```

Comment:
I think we cannot rely on the old proxy contracts. Imagine that 2 years have passed and this generator changed proxies multiple times, for example: X -> Y -> Z. We can not predict what would happen with X proxy in 2 years thus I strongly recommend to avoid querying it. I suggest to make a raw query to Generator's PROXY_REWARD_ASSET and take reward token from there.

Reply:
Fixed